third_party/lua/luadebug.lua

```diff
@@ -413,13 +458,144 @@ local function cmd_eval(code)
+        +100
+--]]
+local function parse_bp(expr)
+    assert(expr)
+    local from, to, file, line = expr:find('^(.-)%s*[+:](%d+)%s*$')
+    if not from then return nil end
+    assert(to == #expr)
+    return file, tonumber(line)
+end
+
+local function _bp_format_expected()
+    dbg_write_error("command expects argument in format filename:NN or " ..
+                    "filename+NN, where NN should be a positive number.")
+end
+
+-- check range to fit in boundaries {from, to}
+local function check_line_max(v, from, to)
+    if type(v) ~= 'number' then
+        dbg_write_error("numeric value expected as line number, " ..
+                        "but received %s", type(v))
+        return false
+    end
+    if v >= from and v <= to then
+        return true
+    end
+    dbg_write_error("line number %d is out of allowed range [%d, %d]",
+                    v, from, to)
+    return false
+end
+
+
+local function cmd_add_breakpoint(bps)
+    assert(bps)
+    local fullfile, line = parse_bp(bps)
+    if not fullfile then
+        _bp_format_expected()
+        return false
+    end
+    if not check_line_max(line, 1, LJ_MAX_LINE) then
+        return false
+    end
+    local info = debug.getinfo(stack_inspect_offset + CMD_STACK_LEVEL, "S")
+    local debuggee = info.source
+    local file = #fullfile > 0 and fullfile or debuggee
+    local normfile = normalize_path(file)
+
+    -- save direct hash (for faster lookup): line -> file
+    if not breakpoints[line] then
+        breakpoints[line] = {}
+    end
+    breakpoints[line][normfile] = true
+    -- save reversed hash information: file -> [lines]
+    if not breakpoints[normfile] then
+        breakpoints[normfile] = {}
+    end
+    breakpoints[normfile][line] = true
+
+    dbg_writeln("Added breakpoint %s:%d.", color_blue(normfile), line)
+    where({source = file, currentline = line}, 0, true)
+    return false
+end
+
+local function _bpd_format_expected()
+    dbg_write_error(" command expects argument specifying breakpoint or" ..
+                    " * for all breakpoints.")
+end
+
+local function cmd_remove_breakpoint(bps)
+    assert(bps)
+    if bps == '*' then
+        breakpoints = {}
+        dbg_writeln("Removed all breakpoints.")
+        return false
+    end
+
+    local fullfile, line = parse_bp(bps)
+    if not fullfile then
+        _bpd_format_expected()
+        return false
+    end
+    if not check_line_max(line, 1, LJ_MAX_LINE) then
+        return false
+    end
+    local info = debug.getinfo(stack_inspect_offset + CMD_STACK_LEVEL, "S")
+    local debuggee = info.source
+    local file = #fullfile > 0 and fullfile or debuggee
+    local normfile = normalize_path(file)
+    local removed = false
+
+    if breakpoints[line] and breakpoints[line][normfile] then
+        breakpoints[line][normfile] = nil
+        removed = true
+    end
+    if breakpoints[file] and breakpoints[file][line] then
+        breakpoints[file][line] = nil
+        removed = true
+    end
```

Comment:
Is it possible that the first check succeeds and the second not? To my understanding `breakpoints[line][file]` and `breakpoints[file][line]` **must** be synced. Maybe assert instead?

Reply:
And why isn't `normfile` used in both cases?

Reply:
> And why isn't normfile used in both cases?
Fixed, thanks!

Reply:
```diff
--- a/third_party/lua/luadebug.lua
+++ b/third_party/lua/luadebug.lua
@@ -581,9 +581,8 @@ local function cmd_remove_breakpoint(bps)
 
     if breakpoints[line] and breakpoints[line][normfile] then
         breakpoints[line][normfile] = nil
-        removed = true
-    end
-    if breakpoints[normfile] and breakpoints[normfile][line] then
+        -- direct and reverse hashes should be in sync
+        assert(breakpoints[normfile] and breakpoints[normfile][line])
         breakpoints[normfile][line] = nil
         removed = true
     end
```